dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -0,0 +1,742 @@
+            output_header.getPositionByName(column.column_name));
+    }
+}
+
+void WindowBlockInputStream::initialWorkspaces()
+{
+    // Initialize window function workspaces.
+    workspaces.reserve(window_description.window_functions_descriptions.size());
+
+    for (const auto & window_function_description : window_description.window_functions_descriptions)
+    {
+        WindowFunctionWorkspace workspace;
+        workspace.window_function = window_function_description.window_function;
+        workspace.column_name = window_function_description.column_name;
+        workspace.argument_columns.assign(window_function_description.argument_names.size(), nullptr);
+        workspace.is_agg_workspace = false;
+        workspaces.push_back(std::move(workspace));
+    }
+    only_have_row_number = onlyHaveRowNumber();
+    only_have_pure_window = onlyHaveRowNumberAndRank();
+}
+
+Block WindowBlockInputStream::readImpl()
+{
+    const auto & stream = children.back();
+    while (!input_is_finished)
+    {
+        if (Block output_block = getOutputBlock())
+        {
+            return output_block;
+        }
+
+        Block block = stream->read();
+        if (!block)
+            input_is_finished = true;
+        appendBlock(block);
+    }
+
+    if (isCancelledOrThrowIfKilled())
+    {
+        return {};
+    }
+
+    // return last partition block, if already return then return null
+    return getOutputBlock();
+}
+
+// Judge whether current_partition_row is end row of partition in current block
+bool WindowBlockInputStream::isDifferentFromPrevPartition(UInt64 current_partition_row)
+{
+    for (size_t i = 0; i < partition_column_indices.size(); ++i)
+    {
+        const auto reference_column
+            = inputAt(prev_frame_start)[partition_column_indices[i]];
+        const auto * compared_column
+            = inputAt(partition_end)[partition_column_indices[i]].get();
+        if (window_description.partition_by[i].collator)
+        {
+            if (compared_column->compareAtWithCollation(current_partition_row,
+                                                        prev_frame_start.row,
+                                                        *reference_column,
+                                                        1 /* nan_direction_hint */,
+                                                        *window_description.partition_by[i].collator)
+                != 0)
+            {
+                return true;
+            }
+        }
+        else
+        {
+            if (compared_column->compareAt(current_partition_row,
+                                           prev_frame_start.row,
+                                           *reference_column,
+                                           1 /* nan_direction_hint */)
+                != 0)
+            {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+void WindowBlockInputStream::advancePartitionEnd()
+{
+    if (partition_ended)
```

Comment:
It looks like `partition_ended` must be false here?
replace by `RUNTIME_ASSERT`?